classes/client.php

```diff
@@ -127,21 +127,27 @@ public static function get_records_without_subscription() {
      * Get client by domain.
      *
      * @param string $domain
+     * @param string $clientid Optional to restrict access to scorm if domain is used by more than one client.
      * @return client
+     * @throws \dml_exception
      */
-    public static function get_record_by_domain(string $domain) {
+    public static function get_record_by_domain(string $domain, string $clientid) {
         global $DB;
 
-        $sql = "SELECT client.*
-                  FROM {scormremote_clients} client
-                 WHERE EXISTS (
-                           SELECT 1
-                             FROM {scormremote_client_domains} dom
-                            WHERE client.id = dom.clientid
-                              AND dom.domain = :domain
-                       )";
+        $clientidclause = ($clientid) ? "AND client.id = :clientid" : "";
 
-        $record = $DB->get_record_sql($sql, ['domain' => $domain]);
+        $sql = "SELECT client.*
+                FROM   mdl_scormremote_clients AS client
+                WHERE  client.id IN (SELECT scd.clientid
+                                     FROM   mdl_scormremote_client_domains AS scd
+                                     WHERE  scd.domain = :domain1
+                                     UNION
+                                     SELECT sc.id AS clientid
+                                     FROM   mdl_scormremote_clients AS sc
+                                     WHERE  sc.primarydomain = :domain2)
+                                     $clientidclause";
+
+        $record = $DB->get_record_sql($sql, ['domain1' => $domain, 'domain2' => $domain, 'clientid' => $clientid]);
```

Comment:
Does this work when $clientid is empty?

I think this could be a lot simpler, if the client id is supplied and valid, then the domain should be ignored and the clients primary domain is used instead.

Lets say we have 2 clients:

acme.com who uses scormcloud.com

and

foobar.cm who also uses scormcloud.com

In this scenario users from both will have a domain of scormcloud.com which isn't what we want, we want to see the real domain which is why we hard code the client id in these cases as otherwise you couldn't tell them apart. Both clients need to have scormcloud in the allowed domain but they will be registered under the primary domain.

Reply:
This function is being called with $origin beign passed as the domain.
I agree that in the example above there is no way to distinguish the two clients. This function will return one of them.
Adding the clientid to this function confirms if the client has the domain in their allowed list. This is in case the domain has been removed from this client (revoked?)
I have left it for now but I this this issue needs to be addressed in another PR.